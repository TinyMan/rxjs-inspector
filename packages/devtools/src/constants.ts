export const enum ConnectionNames {
  PANEL = 'panel',
  CONTENT_SCRIPT = 'content_script',
}

export const enum EventType {
  INIT = 'init',
  NOTIF = 'notif',
}

export const EXTENSION_KEY = '__RXJS_INSPECTOR_EXTENSION__';
