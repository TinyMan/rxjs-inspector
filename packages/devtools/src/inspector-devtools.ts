import { setup, Notif } from '@rxjs-inspector/core';
import { Observable } from 'rxjs';
import { EXTENSION_KEY, EventType } from './constants';
import {
  DevtoolsHook,
  DevtoolsNotifEvent,
  DevtoolsBatchEvent,
  DevtoolsInitEvent,
} from './types';
import { bufferTime } from 'rxjs/operators';
function replaceErrors(key: any, value: any) {
  if (value instanceof Error) {
    let error: { [key: string]: any } = {};

    Object.getOwnPropertyNames(value).forEach(function(key) {
      error[key] = (value as any)[key];
    });

    return error;
  }

  return value;
}

export class InspectorDevtools {
  private notifications$: Observable<Notif> | null = null;
  private extension: DevtoolsHook | null = null;
  constructor(private window: Window & { [EXTENSION_KEY]?: DevtoolsHook }) {
    if (window.hasOwnProperty(EXTENSION_KEY)) {
      this.extension = window[EXTENSION_KEY] as DevtoolsHook;
      this.window.dispatchEvent(
        new CustomEvent(this.extension.namespace, {
          detail: new DevtoolsInitEvent(),
        })
      );

      const { notifications$ } = setup();
      this.notifications$ = notifications$;
      this.notifications$
        .pipe(bufferTime(50))
        .subscribe(buf => buf.length > 0 && this.postBatch(buf));
    }
  }
  private notifToEvent(notif: Notif) {
    // https://stackoverflow.com/questions/28699159/customevent-detail-tainted
    // every object passed through this interface HAVE to be clonable
    return new DevtoolsNotifEvent({
      ...notif,
      value: JSON.stringify(notif.value, replaceErrors),
      observable: undefined,
    });
  }
  private postBatch(batch: Notif[]) {
    if (this.extension) {
      const event = new CustomEvent(this.extension.namespace, {
        detail: new DevtoolsBatchEvent(batch.map(n => this.notifToEvent(n))),
      });
      this.window.dispatchEvent(event);
    }
  }
}
