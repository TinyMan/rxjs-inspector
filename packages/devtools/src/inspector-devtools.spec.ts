import { InspectorDevtools } from './inspector-devtools';
import { Observable } from 'rxjs';
import { EXTENSION_KEY } from './constants';

describe('InspectorDevtools', () => {
  let proto: typeof Observable.prototype;
  beforeEach(() => {
    proto = Object.assign({}, Observable.prototype);
    Object.defineProperty(window, EXTENSION_KEY, {
      value: {
        namespace: 'test-namespace',
      },
      configurable: true,
    });
  });
  afterEach(() => {
    Object.assign(Observable.prototype, {}, proto);
    Object.defineProperty(window, EXTENSION_KEY, {
      value: undefined,
      configurable: true,
    });
  });
  it('should construct', () => {
    expect(() => new InspectorDevtools(window)).not.toThrow();
  });
  test('should send a init message to the extension', () => {
    const dispatchEvent = jest.fn();
    Object.defineProperty(window, 'dispatchEvent', {
      value: dispatchEvent,
    });
    new InspectorDevtools(window);
    expect(dispatchEvent).toHaveBeenCalledTimes(1);
    expect(dispatchEvent.mock.calls[0][0].detail).toMatchSnapshot();
  });
});
