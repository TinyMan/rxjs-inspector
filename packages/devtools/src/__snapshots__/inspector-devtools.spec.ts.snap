// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InspectorDevtools should send a init message to the extension 1`] = `
Object {
  "type": "init",
}
`;
